Interpolation Surrogate

pbauman opened the pull request:
Adds linear Lagrange interpolant surrogate for arbitrary dimension. Note I've defined a hierarchy of classes to promote an interface and reusability. In particular, I intended to create subclasses of the Gaussian likelihoods, for example, that take a pointer to `SurrogateBase`, then it becomes that much easier for the user to use surrogate models in their likelihood. I'd appreciate feedback on this point, particularly since there's some GP work underway already --- we'd want GP model surrogates to subclass this base. Future spectral projection surrogates would also subclass `SurrogateBase`.

@dmcdougall I'd appreciate you taking a good look through the comments and make sure that they make sense and you can follow the code. The equations to code translation may not be obvious everywhere here and I want to make sure the documentation is clear.

I still plan on adding a `Generator` (other name suggestions welcome) that will let the user supply their model and then QUESO can farm off filling in the values over subenvironments (in the QUESO parlance) over many processors. Also want to add constructors that take objects that will populate the surrogate. But at least this can start getting a review.

I have unit tests out to 4D. This won't be of much utility beyond that number of dimensions I don't think, but at least gives some warm and fuzzies about the correctness.

pbauman commented:
Rebased on dev. (serial) Tests and example added. I'm just about ready. There's a bug in the parallel case that I'm chasing, but once that's resolved, this is good to go. Criticism welcome, particularly about documentation. I'm hoping I commented enough that this is self-documenting, but let me know if it's not.

pbauman commented:
OK, I believe this is good to go. Figured out the parallel bug and the examples works for me on 1-8 subenviroments.

dmcdougall commented:
You need a [small modification to this variable here](https://github.com/libqueso/queso/blob/dev/doxygen/queso.dox.in#L568) in `queso.dox` so that your documentation will get rendered automatically by the build system.

dmcdougall commented on src/surrogates/inc/SurrogateBase.h:
```diff
@@ -0,0 +1,52 @@
+    virtual ~SurrogateBase(){};
+
+    //! Method to return value given the parameter vector
+    virtual double evaluate(const V & domainVector) const =0;
```
This looks as though we're only supporting surrogates of models that map `\mathbb{R}^n \to \mathbb{R}`, is that correct?  I.e., we're not currently supporting `\mathbb{R}^n \to \mathbb{R}^m`

I'm not suggesting implementing the multi-dimensional output case in this PR (if at all), I'm merely asking for purposes pertaining to my understanding.

pbauman replied:
Correct. If the user has multiple outputs, I envision creating one of these for each output. I'll update the documentation accordingly.

dmcdougall commented on src/surrogates/inc/InterpolationSurrogateBase.h:
```diff
@@ -0,0 +1,69 @@
+#include <queso/SurrogateBase.h>
+#include <queso/BoxSubset.h>
+#include <queso/VectorSpace.h>
+#include <queso/InterpolationSurrogateData.h>
```
This is up to you, but it looks like the only header you need here is the base class header.  You can get away with just a forward declaration of `InterpolationSurrogateData`.

pbauman replied:
Good call. Thanks for catching my slop.

dmcdougall commented on src/surrogates/inc/InterpolationSurrogateBase.h:
```diff
@@ -0,0 +1,69 @@
+  //! Base class for interpolation-based surrogates
+  /*! This class is used for surrogoate approximations of a model using interpolation.
+      Subclasses will define behavior of interpolant, but common assumptions
+      are: 1. Bounded domain; future work may extend behavior to unbounded domains.
```
This list doesn't render properly in doxygen.

pbauman replied:
:-/ OK, thanks, I'll take a look.

dmcdougall commented on src/surrogates/inc/InterpolationSurrogateBase.h:
```diff
@@ -0,0 +1,69 @@
+namespace QUESO
+{
+  //! Base class for interpolation-based surrogates
+  /*! This class is used for surrogoate approximations of a model using interpolation.
```
Perhaps clarify the vector spaces the model maps from/to.

pbauman replied:
Hmm. I wonder if maybe the base class should be renamed to reflect that this is designed for single quantity output (vs. multiple quantity)?

dmcdougall commented:
@pbauman I've looked through the documentation and design and I think this looks pretty solid.  I've still to go through and give you some more fine-grained feedback and things I need clarification on.

Thanks for taking the time to put this together; first impressions are it'll be very useful for small dimensional problems.

pbauman commented:
>I've still to go through and give you some more fine-grained feedback and things I need clarification on.

Thanks for taking a look. I'll push a commit addressing the comments you made.

> first impressions are it'll be very useful for small dimensional problems.

Yeah, I can't see this really being useful beyond 4, maybe 5 dimensions, but if your science has 3 or less parameters, or even just to play with the problem a bit, this helps get you started. I plan to use this on the nitridation problem ASAP.

Future work will build around `SurrogateBase`, e.g. spectral projection (like we discussed at SIAM). Would like GPs, but I'm still not clear exactly what's going on there. I gather @roystgnr is working on GPMSA, but not sure of anything beyond that.

pbauman commented:
Rebased and address @dmcdougall's current comments.

dmcdougall commented on src/surrogates/inc/InterpolationSurrogateBase.h:
```diff
@@ -0,0 +1,74 @@
+  public:
+
+    //! Constructor
+    /*! n_points should be a vector with dimension matching the dimension of
```
`data.n_points` instead of `n_points`?

dmcdougall commented on src/surrogates/inc/LinearLagrangeInterpolationSurrogate.h:
```diff
@@ -0,0 +1,110 @@
+    void compute_interval_indices(const V & domainVector,
+                                  std::vector<unsigned int>& indices) const;
+
+    //! Helper function to populate bounding values for the intervales in each dimension
```
Typo: intervales.

dmcdougall commented:
This is well documented, well tested and well written code.  Good job.

I've gone through the documentation and barring a few minor typos everything looks good.

I've gone through the code itself.  I'm not going to comment on or correct the code style because the code style throughout `queso` itself is not consistent.  I have some plans to amend in the future.  There are plenty of comments which make this a pleasure to read.

I've gone through the tests and I'm very happy to the extent this has been tested.  Well done.

I've also build the doxygen documentation and looked at how the classes are structured and what interface we're exposing to the user.  You've done a good job in making your interfaces extensible and leave everything open-ended for changing requirements and new surrogate functionality in the future.  At first I was going to suggest subclassing `BaseScalarFunction` for the `evaluate_model` method, but actually I think not doing that is more appropriate; I was confusing myself between *model* evaluations and *likelihood* evaluations.  In fact, you take a very similar approach to the one I took in coding up the canned Gaussian likelihood classes a while back.

This is a very big thumbs-up from me.  Thank you so much for taking the time and effort to put this together.

@pbauman Fix up the minor typos and don't forget to make a note in the changelog.  After you've done that I will merge this.

P.S.  Thank you for using the `queso_assert*` stuff.  Code is much more readable than using the old `UQ_FATAL_MACRO*` macros.

pbauman commented:
Fixed. Thanks for the review. I need to add a reader, writer, and checkpointing, but I'll do that in a separate PR.

pbauman commented:
Rebased
